packages/cli/src/fingerprint/fingerprintWatchCommand.js

```diff
@@ -7,29 +9,51 @@ class FingerprintWatchCommand {
     this.directory = directory;
     this.print = true;
     this.numProcessed = 0;
+    this.pidfilePath =
+      process.env.APPMAP_WRITE_PIDFILE &&
+      path.join(this.directory, 'index.pid');
   }
 
-  execute() {
-    if (verbose()) {
-      console.warn(`Watching appmaps in ${this.directory}`);
+  removePidfile() {
+    if (this.pidfilePath) {
+      fs.removeSync(this.pidfilePath);
     }
+  }
 
-    this.fpQueue = new FingerprintQueue();
-    this.fpQueue.setCounterFn(() => {
-      this.numProcessed += 1;
-    });
-    this.fpQueue.process();
-    this.watcher = chokidar.watch(`${this.directory}/**/*.appmap.json`, {
-      ignoreInitial: true,
+  execute() {
+    return new Promise((resolve) => {
+      this.fpQueue = new FingerprintQueue();
+      this.fpQueue.setCounterFn(() => {
+        this.numProcessed += 1;
+      });
+      this.fpQueue.process();
+      this.watcher = chokidar.watch(`${this.directory}/**/*.appmap.json`, {
+        ignoreInitial: true,
+      });
+      const { pidfilePath, directory } = this;
+      const removePidfile = this.removePidfile.bind(this);
+      this.watcher
+        .on('add', this.added.bind(this))
+        .on('change', this.changed.bind(this))
+        .on('unlink', this.removed.bind(this))
+        .on('ready', () => {
+          if (pidfilePath) {
+            fs.outputFileSync(pidfilePath, `${process.pid}`);
+            process.on('exit', removePidfile);
```

Comment:
Does this function need to be bound to this?

Reply:
You mean the arrow function being passed to `.on('ready'`, right? It does not need to be bound to `this`, because it's getting all the variables it needs from lines 33 & 34 above:
```
      const { pidfilePath, directory } = this;
      const removePidfile = this.removePidfile.bind(this);
```
I did this because of the way the functions are nested. But, looking at it again, it probably makes sense to factor the whole thing out into a separate method. I'll do that.